### PR TITLE
Potential changes to Rules for AI Triage Service

### PR DESCRIPTION
**Pull request by jeo02**
Using this PR to discuss changes to the AI Label Service to include AI comments. Changes mainly surround the response structure, and how we react to the new fields.

```
{
    "labels" : string[],  // zero or two
    "suggestion" : string | null,  // never empty
    "solution" : string | null  // never empty
}
```

### CODE REVIEW

File: tools/github-event-processor/RULES.md [hunk 1]
```diff
@@ -70,10 +70,12 @@ _**Predictions made**_
...
 ```json
 {
-  "labels": [
+  "labels" : [
     "Storage",
     "Client"
-  ]
+  ],  
+  "suggestion" : string | null,  
+  "solution" : string | null
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:78]:
```suggestion
  "suggestion" : null,  
  "solution" : null  
```

File: tools/github-event-processor/RULES.md [hunk 2]
```diff
@@ -70,10 +70,12 @@ _**Predictions made**_
 
 ```json
 {
-  "labels": [
+  "labels" : [
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:73]: I'd suggest that we break this out into the 4 possible states with examples for each:

- Manual triage needed.   (empty labels, no suggestions, no solution)
- Label prediction made for triage.  (labels populated, no suggestion, no solution)
- Label prediction with suggestion (labels populated, suggestion populated, no solution)
- Label prediction with solution (labels populated, no suggestion, solution populated)

Comment (jeo02) [reply tools/github-event-processor/RULES.md:73]: For suggestion/solution example I just provided the first bit. Not sure if you would want a complete example.

File: tools/github-event-processor/RULES.md [hunk 3]
```diff
@@ -133,6 +164,12 @@ IF labels were predicted:
     IF "needs-team-triage" is not being added to the issue
          - Add "needs-team-attention" label to the issue
 
+    IF suggestions is populated
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:167]:
```suggestion
    IF suggestion is populated
```

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/GitHubEventClient.cs [hunk 4]
```diff
@@ -1321,23 +1321,18 @@ public virtual RulesConfiguration LoadRulesConfiguration(string rulesConfigLocat
...
-        /// <summary>
-        /// Class to hold the response from the AI Label Service query
-        /// </summary>
-        public class LabelResponse
+        public virtual async Task<IssueTriageOutput> QueryAILabelService(IssueEventGitHubPayload issueEventPayload)
         {
-            public string[] Labels { get; set; }
-        }
+            //Empty response for failures
+            IssueTriageOutput empty = new IssueTriageOutput { Labels = [], Suggestion = null, Solution = null };
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/GitHubEventClient.cs:1327]: We should make this a static method on `IssueTriageOutput` so that we can do  `IssueTriageOutput.Empty` here.

**Comment by jeo02** [comment]
Changed the output structure to be more general 
```
{
    "labels" : ["service", "category"],
    "answer" : "Hi. I'm AI...",
    "answerType" : "none" | "sugestion" | "solution"
}
```

File: tools/github-event-processor/RULES.md [hunk 5]
```diff
@@ -110,8 +110,8 @@ _**No predictions**_
 ```json
 {
   "labels": [],  
-  "suggestion" : null,  
-  "solution" : null  
+  "answer" : null,  
+  "answerType" : "none"
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:114]: Since we're not doing an enum, should we go with `null` here or use the "none" string?    I'm on the fence and don't feel strongly either way.

File: tools/github-event-processor/RULES.md [hunk 6]
```diff
@@ -48,8 +48,8 @@
...
-(<https://github.com/Azure/azure-sdk-tools/blob/main/tools/github-labels/repositories.txt>). For libraries that are not supported it an empty response will be returned.
+This is a REST API that evaluates the content of an issue and attempts to predict a set of labels and provide a relevant comment. If predictions is made with an extreme level of confidence, the service will return one service label, one category label and a solution comment. In this case an `issue-addressed` is added to the issue to close it. If a prediction is made with a reasonable level of confidence, the service will return one service label, one category label, and a suggestion comment.  If confidence thresholds were not met, labels nor a comment are provided. The AI label service can only a response with labels for [officially supported repositories]
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:51]:
```suggestion
This is a REST API that evaluates the content of an issue and attempts to predict a set of labels and provide a relevant answer. If predictions are made with a high degree of confidence, the service will return one service label, one category label and a solution comment. In this case an `issue-addressed` is added to the issue to close it. If a prediction is made with a reasonable level of confidence, the service will return one service label, one category label, and a suggestion comment.  If confidence thresholds were not met, labels nor a comment are provided. The AI label service can only a response with labels for [officially supported repositories]
```

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/EventProcessing/IssueProcessing.cs [hunk 7]
```diff
@@ -178,7 +178,7 @@ public static async Task InitialIssueTriage(GitHubEventClient gitHubEventClient,
                                     }
                                 }
                                 // If the issue had a valid assignee and only labels provided add the comment
-                                if (hasValidAssignee && triageOutput.Suggestion == null && triageOutput.Solution == null)
+                                if (hasValidAssignee && triageOutput.AnswerType == "none" && triageOutput.Labels.Count > 0)
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/EventProcessing/IssueProcessing.cs:181]: I'd add a `string.IsNullOrEmpty` guard for "AnswerType" here too - just to be safe.

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs [hunk 8]
```diff
@@ -507,13 +507,13 @@ public async Task TestInitialIssueTriage(string rule,
...
             }
 
-            if (AIServiceSuggestion != null)
+            if (AIServiceAnswer != null)
             {
-                mockGitHubEventClient.AIServiceSuggestion = AIServiceSuggestion;
+                mockGitHubEventClient.AIServiceAnswer = AIServiceAnswer;
             }
-            if (AIServiceSolution != null)
+            if (AIServiceAnswerType != null)
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs:514]: I'd avoid making this assumption in the tests and use an explicit "none" if that's what we're expecting.   We should probably also have a test case for `null` and for "".

File: tools/github-event-processor/RULES.md (hunk 5 above)
Comment (jeo02) [reply tools/github-event-processor/RULES.md:114]: think null would make more sense for consistency's sake (since answer is also null when nothings there).

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs (hunk 8 above)
Comment (jeo02) [reply tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs:514]: Changed to be null so should not be a problem but will be adding a test for empty

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs [hunk 9]
```diff
@@ -42,6 +42,9 @@ public class MockGitHubEventClient: GitHubEventClient
         public List<PullRequestFile> PullRequestFiles { get; set; } = new List<PullRequestFile>();
 
         public List<string> AILabelServiceReturn { get; set; } = new List<string>();
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs:44]:
```suggestion
        public List<string> AIServiceReturn { get; set; } = new List<string>();
```

nittiest of nits, but we should align the names between the service-related variables.

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/Utils/IssueTriageOutput.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+{
+    /// <summary>
+    /// Class used to deserialize AI Triage output.
+    /// Has four states:
+    /// - Empty: No labels, no answer, no answer type
+    /// - Labels: Labels are provided, no answer, no answer type. 
+    /// - Suggestion: Labels are provided, answer is provided, answer type = "suggestion".
+    /// - Solution: Labels are provided, answer is provided, answer type = "solution".
+    /// </summary>
+    public class IssueTriageOutput
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor/Utils/IssueTriageOutput.cs:17]:
```suggestion
    public class IssueTriageResponse
```

another really nitty thing.   Generally the name "Response" is preferred in across the .NET packages for "the service returned this thinger."    It doesn't really matter here, as this is internal, but felt worth mentioning.

I leave it to your discretion whether to rename or ignore me.   I'm good either way.  😄

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs [hunk 11]
```diff
@@ -26,14 +26,15 @@ public class IssueProcessingTests : ProcessingTestBase
         // 6. RuleState ruleStateServiceAttention - Whether or not the ServiceAttention rule is on or off
         //
         // CODEOWNERS changes necessary to test
-        // The CODEOWNERS file needs entries that match the AIReturnLabels, in theory the AI Label Service will return one
+        // The CODEOWNERS file needs entries that match the AIReturnLabels, in theory the AI Issue Triage will return one
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/Static/IssueProcessingTests.cs:29]:
```suggestion
        // The CODEOWNERS file needs entries that match the AIReturnLabels, in theory the AI Issue Triage service will return one
```

nit:  We're inconsistent in using "service" vs. "Service" vs. "" in these.   Doesn't matter which we choose, but it would be nice to align them.

File: tools/github-event-processor/RULES.md [hunk 12]
```diff
@@ -66,22 +65,52 @@ This is a REST API that evaluates the content of an issue and attempts to predic
 
 #### Example responses
 
-_**Predictions made**_
+_**Only Labels Predicted**_
```
Comment (jsquire) [inline tools/github-event-processor/RULES.md:68]:
```suggestion
_**Only labels predicted**_
```

**Review by jsquire — COMMENTED** [review]
The important bits look great.   I left a couple of really nitty thoughts.   When you're ready to merge, let me know and I'll do one final pass.

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs [hunk 13]
```diff
@@ -42,6 +42,9 @@ public class MockGitHubEventClient: GitHubEventClient
...
+        public string AIServiceAnswer { get; set; } = null;
+        public string AIServiceAnswerType { get; set; } = null;
```
Comment (benbp) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs:46]:
```suggestion
        public string? AIServiceAnswer { get; set; } = null;
        public string? AIServiceAnswerType { get; set; } = null;
```

Comment (benbp) [reply tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs:46]: This is probably why you are getting a compile failure in the tests.

Comment (jeo02) [reply tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs]: Thanks Ben!!

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs (hunk 9 above)
Comment (jeo02) [reply tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs]: Changed it to `AIServiceLabels`

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs [hunk 14]
```diff
@@ -469,10 +472,14 @@ public override RulesConfiguration LoadRulesConfiguration(string? rulesConfigLoc
         // The mock won't be calling the actual service and since the method it's overriding is
         // async, the warning needs to be disabled.
 #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
-        public override async Task<List<string>> QueryAILabelService(IssueEventGitHubPayload issueEventPayload)
+        public override async Task<IssueTriageResponse> QueryAILabelService(IssueEventGitHubPayload issueEventPayload)
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs:475]:
```suggestion
        public override Task<IssueTriageResponse> QueryAILabelService(IssueEventGitHubPayload issueEventPayload)
```

You only need the `async` keyword if you use `await`.   This will let you remove the warning suppression.

File: tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs [hunk 15]
```diff
@@ -469,10 +472,14 @@ public override RulesConfiguration LoadRulesConfiguration(string? rulesConfigLoc
...
 #pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
         {
-            return AILabelServiceReturn;
+            return new IssueTriageResponse {
```
Comment (jsquire) [inline tools/github-event-processor/Azure.Sdk.Tools.GitHubEventProcessor.Tests/MockGitHubEventClient.cs:478]:
```suggestion
            return Task.FromResult(
                new IssueTriageResponse { 
```

You'll have to close the parenthesis too...

**Review by jsquire — APPROVED** [review]
